frameworks/modular/module.h

```diff
@@ -0,0 +1,296 @@
+/* Copyright (c) 2017, ARM Limited. All rights reserved.
+ *
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * Modular framework solves the problem of choosing a module
+ * from multiple modules of a subsystem.
+ *
+ * The choice is available during compile time or during runtime
+ * or initialization. The runtime choice could be using shared
+ * libraries or dynamic loadable libraries.
+ *
+ * Multiple modules of the same subsystem can be built into
+ * individual static libraries(.a), shared libraries(.so) to be
+ * dynamically linked or loaded, and use constructor functions
+ * to register themselves.
+ *
+ * A subsystem can choose one active module and provide APIs to
+ * switch between modules.
+ *
+ * Alternatively, subsystem can load multiple modules and
+ * determine the APIs route in runtime.
+ *
+ * In order to gain full possible performance, the subsystem
+ * allows for choosing a specific module at compile time.
+ * This eliminates the need to choose the module using function
+ * pointer table.
+ *
+ * This framework tries to minimizes dependencies to the linked
+ * list and rwlock facilities only.
+ */
+
+#ifndef MODULE_H
+#define MODULE_H
+
+#include <stdbool.h>
+#include <odp/api/rwlock.h>
+#include "list.h"
+
+/* Forward declaration */
+typedef struct odp_module_base odp_module_base_t;
+
+/* Subsystem */
+typedef struct {
+	odp_rwlock_t lock;
+	uint32_t version;
+	const char *name;
```

Comment:
Yes, since we will define the subsystem with below macro, name will remains valid throughout the lifetime of the struct (string paste):

```C
/* Define an ODP subsystem in source file */
#define ODP_SUBSYSTEM_DEFINE(_name, _description, _version)	\
	odp_subsystem_t __subsystem(_name) =			\
	{							\
		.lock = ODP_RWLOCK_UNLOCKED,			\
		.name = # _name,				\
		.version = _version,				\
		.description = _description,			\
	}
```